cpp/DEVELOPER_GUIDE.md

```diff
@@ -5,9 +5,22 @@ This document summarizes rules and best practices for contributions to the cuML
 Please start by reading [CONTRIBUTING.md](../CONTRIBUTING.md).
 
 ## Thread safety
-cuML is thread safe so its functions can be called from multiple host threads if they use different handles.
 
-The implementation of cuML is single threaded.
+With the exception of the cumlHandle, cuML algorithms should maintain thread-safety and are, in general, 
+assumed to be single threaded. This means they should be able to be called from multiple host threads so 
+long as different instances of `cumlHandle` are used.
+
+Exceptions are made for algorithms that can take advantage of multiple CUDA streams in order to oversubscribe 
+or increase occupancy on a single GPU. In these cases, the use of multiple host threads within cuML algorithms
+should be used only to maintain concurrency of the underlying CUDA streams. Multiple host threads should be 
+used sparingly, be bounded, and should steer clear of performing CPU-intensive computations.
+
+A good example of an acceptable use of host threads can be found in [this blog article](https://devblogs.nvidia.com/gpu-pro-tip-cuda-7-streams-simplify-concurrency/)
+
+Though cuML's build enables OpenMP by default, cuML algorithms should still function properly even when OpenMP has been
+disabled. For this reason, it would be better to utilize OpenMP rather than raw pthreads for spawning host threads. 
```

Comment:
This is a little unclear. If you use OpenMP rather than pthreads, will it function properly with OpenMP disabled?

Reply:
Agree with @harrism this is unclear. I would suggest the following: "To avoid compatibility issues between different threading models the only multi threading programming allowed in cuML is OpenMP."